pkg/gadget-service/service.go

```diff
@@ -0,0 +1,291 @@
+		logger.Errorf("gadget not found: %s/%s", request.GadgetCategory, request.GadgetName)
+		return nil
+	}
+
+	// TODO: Remove
+	logger.Debugf("Params")
+	for k, v := range request.Params {
+		logger.Debugf("- %s: %q", k, v)
+	}
+
+	// Initialize Operators
+	err = operators.GetAll().Init(operators.GlobalParamsCollection())
+	if err != nil {
+		logger.Errorf("initialize operators: %v", err)
+		return err
+	}
+
+	ops := operators.GetOperatorsForGadget(gadgetDesc)
+
+	operatorParams := ops.ParamCollection()
+	err = operatorParams.CopyFromMap(request.Params, "operator.")
+	if err != nil {
+		logger.Errorf("setting operator parameters: %v", err)
+		return nil
+	}
+
+	parser := gadgetDesc.Parser()
+
+	runtimeParams := runtime.ParamDescs().ToParams()
+	err = runtimeParams.CopyFromMap(request.Params, "runtime.")
+	if err != nil {
+		logger.Errorf("setting runtime parameters: %v", err)
+		return nil
+	}
+
+	gadgetParamDescs := gadgetDesc.ParamDescs()
+	gadgetParamDescs.Add(gadgets.GadgetParams(gadgetDesc, parser)...)
+	gadgetParams := gadgetParamDescs.ToParams()
+	err = gadgetParams.CopyFromMap(request.Params, "")
+	if err != nil {
+		logger.Errorf("setting gadget parameters: %v", err)
+		return nil
+	}
+
+	// Create payload buffer
+	outputBuffer := make(chan *pb.GadgetEvent, 1024) // TODO: Discuss 1024
+	outputDone := make(chan bool)
+	defer func() {
+		outputDone <- true
+	}()
+	seq := uint32(0)
+	var seqLock sync.Mutex
+
+	if parser != nil {
+		parser.SetLogCallback(logger.Logf)
+		parser.SetEventCallback(func(ev any) {
+			// Marshal messages to JSON
+			// Normally, it would be better to have this in the pump below rather than marshaling events that
+			// would be dropped anyway. However, we're optimistic that this occurs rarely and instead prevent using
+			// ev in another thread.
+			data, _ := json.Marshal(ev)
+			lev := &pb.GadgetEvent{
+				Type:    pb.EventTypeGadgetPayload,
+				Payload: data,
+			}
+
+			seqLock.Lock()
+			seq++
+			lev.Seq = seq
+			select {
+			case outputBuffer <- lev:
+			default:
+			}
+			seqLock.Unlock()
+		})
+	}
+
+	go func() {
+		// Message pump to handle slow readers
+		for {
+			select {
+			case ev := <-outputBuffer:
+				runGadget.Send(ev)
+			case <-outputDone:
+				return
+			}
+		}
+	}()
+
+	// Assign a unique ID - this will be used in the future
+	runID := uuid.New().String()
+
+	// Send Job ID to client
+	err = runGadget.Send(&pb.GadgetEvent{
+		Type:    pb.EventTypeGadgetJobID,
+		Payload: []byte(runID),
+	})
+	if err != nil {
+		logger.Warnf("sending JobID: %v", err)
+		return nil
+	}
+
+	// Create new Gadget Context
+	gadgetCtx := gadgetcontext.New(
+		ctx,
+		runID,
+		runtime,
+		runtimeParams,
+		gadgetDesc,
+		gadgetParams,
+		operatorParams,
+		parser,
+		logger,
+		time.Duration(request.Timeout),
+	)
+
+	// Handle commands sent by the client
+	go func() {
+		defer func() {
+			logger.Debugf("runner exited")
+		}()
+		for {
+			msg, err := runGadget.Recv()
+			if err != nil {
+				return
+			}
```

Comment:
We should at least notify the client about this error or, at least, log it. I mean, this failure means we received a request and we failed to read it and we will not listen for other requests anymore. I'd say it is an important error.

Reply:
AFAIK, it's much more likely that the connection has been closed when that function returns an error - and then there's no point in trying to inform the client about it :) But it certainly should call `cancel()` in this case and maybe log to the server logs.

The slight chance of there being a communication problem (gRPC protocol) should be handled just like a disconnection IMHO. The protocol would probably be in an undefined state at that place.